site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+
+In this vein, we have added a web animations meta provider in Beta 4.  It provides a way to easily animate DOM nodes in a reactive way, as just part of your render.  For example:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+import WebAnimation from '@dojo/widget-core/meta/WebAnimation';
+
+export default class AnimatedWidget extends WidgetBase {
+    protected render() {
+        const animate = {
+            id: 'rootAnimation',
+            effects: [
+                { height: '10px' },
+                { height: '100px' }
+            ],
+            controls: {
+                play: true
+            }
+        };
+
+        this.meta(WebAnimation).animate('root', animate);
+
+        return v('div', {
+            key: 'root'
+        });
+    }
+}
+```
+
+## Custom Elements/Web Components
+
+One of the main objectives for Dojo 2 was that there was a high level of interoperability with Web Components, both using Web Components in Dojo 2 applications but also being able to use Dojo 2 widgets easily as Custom Elements.  We made several changes to the widgeting system to better support Dojo 2 widgets used as Custom Elements, specifically making it efficient to use children Custom Elements of a Dojo 2 widget Custom Element.
+
+So what you would write in TypeScript:
+
+```typescript
+class Example extends WidgetBase {
+
+  private _onSelected(data: any) {
+    console.log(data);
+  }
+
+  protected render() {
+    return w(Menu, { onSelected: this._onSelected }, [
+      w(MenuItem, { key: 'a', title: 'Menu Item A' }),
+      w(MenuItem, { key: 'b', title: 'Menu Item B', selected: true }),
+      w(MenuItem, { key: 'c', title: 'Menu Item C' })
+    ]);
+  }
+}
+```
+
+Can be written as this using when using Custom Elements:
+
+```html
+<demo-menu id="menu">
+  <demo-menu-item id="a" title="Menu Item A"></demo-menu-item>
+  <demo-menu-item id="b" title="Menu Item B"></demo-menu-item>
+  <demo-menu-item id="c" title="Menu Item C" selected></demo-menu-item>
+</demo-menu>
+<script>
+  a.data = { foo: 'bar' };
+  menu.addEventListener('selected', function(event) {
+    const detail = event.detail;
+    console.log(detail);
+  });
+</script>
+```
+
+We have an [live example](https://dojo.github.io/examples/custom-element-menu/) of this, along with the [source code on GitHub](https://github.com/dojo/examples/tree/master/custom-element-menu).
+
+## Other widget system improvements
+
+There were other changes to the widgeting system which enable some features and capabilities we think can make widgets more robust:
+
+* The way event listeners behave on virtual DOM nodes has changed, and we now allow the listeners to change between renders.  Previously this would have errored.
```

Comment:
The way event listeners behave on virtual DOM nodes has changed, and we now allow the listeners to change between renders.  Previously this would have errored. -> Event listeners on virtual DOM nodes now allow the listeners to change between renders.